### PR TITLE
clustermesh: Add an infrastructure for connect time parameter exchange and capability negotiation

### PR DESCRIPTION
**Pull request by YutaroHayakawa**
Add a new kvstore object `CiliumClusterConfig` with path `cilium/cluster-config/<cluster name>`. The object is created per cluster by clustermesh-apiserver and obtained by cilium-agent at start-up time. The use cases of the objects are

**Getting per-cluster parameter**
ClusterMesh with overlapping PodCIDR support needs this for getting ClusterID on connect time and setup kvstore observers to annotate incoming objects with remote ClusterID. We can put such parameters to `CiliumClusterConfig`.

**Capability negotiation and configuration validation**
When there's a capability mismatch or configuration mismatch between clusters, we may not be able to connect those clusters correctly. Currently, we don't have any mechanism to check such a mismatch. With `CiliumClusterConfig`, we can check that on connect time. Currently, we only check ClusterID duplication but can add any capabilities/configuration in the future.

```release-note
clustermesh: Add an infrastructure to connect time parameter exchange and capability negotiation
```

### CODE REVIEW

File: pkg/clustermesh/clustermesh.go [hunk 1]
```diff
@@ -73,6 +76,43 @@ type Configuration struct {
...
+
+	val, err := json.Marshal(config)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+	defer cancel()
+
+	_, err = kvstore.Client().UpdateIfDifferent(ctx, key, val, false)
```
Comment (aanm) [inline pkg/clustermesh/clustermesh.go]: Shouldn't we update with a "lease"? This way the configuration would be removed automatically once the lease expired.

File: pkg/clustermesh/remote_cluster.go [hunk 2]
```diff
@@ -177,6 +184,21 @@ func (rc *remoteCluster) restartRemoteConnection(allocator RemoteIdentityWatcher
 
 				rc.getLogger().Info("Connection to remote cluster established")
 
+				config, err := GetClusterConfig(rc.name, backend)
+				if err != nil && config == nil {
+					rc.getLogger().Warning("Remote cluster doesn't have cluster configuration, falling back to the old behavior")
```
Comment (aanm) [inline pkg/clustermesh/remote_cluster.go]: Let's add one more sentence explaining that to the user that this is expected when performing an upgrade where the old version doesn't have this new behavior implemented.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/clustermesh/clustermesh.go (hunk 1 above)
Comment (YutaroHayakawa) [reply pkg/clustermesh/clustermesh.go]: Oops, thank you. I fixed.

File: pkg/clustermesh/remote_cluster.go (hunk 2 above)
Comment (YutaroHayakawa) [reply pkg/clustermesh/remote_cluster.go]: Yep, good idea. I added.

**Review by aanm — APPROVED** [review]
LGTM @YutaroHayakawa just a heads-up that since we have branched v1.13, this PR will not be part of the 1.13 unless it is set with the needs-backport/1.13 label.

**Comment by YutaroHayakawa** [comment]
Ok, let me put it. Thanks for the heads up.

**Comment by YutaroHayakawa** [comment]
Hitting known flakes
ci-l4lb: cilium-config-agent issue, need rebasing
ci-aks: cilium-config-agent issue, need rebasing
ci-eks: cilium-config-agent issue, need rebasing
ci-gke: https://github.com/cilium/cilium/issues/21652 (GKE IP exhaustion)
ci-multicluster: https://github.com/cilium/cilium/issues/21652 (GKE IP exhaustion)

**Comment by YutaroHayakawa** [comment]
Failed tests
ci-aks: Seems like a new flake. Filed an issue https://github.com/cilium/cilium/issues/22681. See if retrying help.

**Comment by YutaroHayakawa** [comment]
All tests are passed. Making this ready-to-merge.

**Comment by pchaigno** [comment]
Should @aanm become a member of @cilium/sig-clustermesh?

**Pull request merged**